### PR TITLE
AVM performance

### PR DESCRIPTION
**Pull request by StephenButtolph**
- Added support for xput tests on the AVM
- Implemented an AVM wallet for throughput tests.
- Fixed credential bug in the AVM for transactions that depend on unconfirmed UTXOs.

### CODE REVIEW

File: vms/avm/vm.go [hunk 1]
```diff
@@ -252,7 +252,7 @@ func (vm *VM) GetTx(txID ids.ID) (snowstorm.Tx, error) {
  */
 
 // IssueTx attempts to send a transaction to consensus
-func (vm *VM) IssueTx(b []byte) (ids.ID, error) {
+func (vm *VM) IssueTx(b []byte, onDecide func(choices.Status)) (ids.ID, error) {
```
Comment (danlaine) [inline vms/avm/vm.go]: Can you put a comment specifying what `onDecide` is?

File: vms/avm/vm_test.go [hunk 2]
```diff
@@ -547,3 +547,159 @@ func TestGenesisGetUTXOs(t *testing.T) {
 		t.Fatalf("Wrong number of utxos (%d) returned", len(utxos))
 	}
 }
+
+func TestIssueDependentTx(t *testing.T) {
```
Comment (danlaine) [inline vms/avm/vm_test.go]: Can you add a comment specifying what this is testing?

File: vms/spchainvm/key_chain.go [hunk 3]
```diff
@@ -28,16 +28,16 @@ type KeyChain struct {
...
 		keyMap:  make(map[[20]byte]int),
 	}
 }
 
 // New returns a newly generated private key
-func (kc *KeyChain) New() *crypto.PrivateKeySECP256K1R {
+func (kc *Keychain) New() *crypto.PrivateKeySECP256K1R {
 	factory := &crypto.FactorySECP256K1R{}
 
 	skGen, _ := factory.NewPrivateKey()
```
Comment (danlaine) [inline vms/spchainvm/key_chain.go]: Should keychain have a log so we can log errors in places like this?

File: vms/avm/vm.go (hunk 1 above)
Comment (StephenButtolph) [reply vms/avm/vm.go]: Updated

File: vms/avm/vm_test.go (hunk 2 above)
Comment (StephenButtolph) [reply vms/avm/vm_test.go]: Updated

File: vms/spchainvm/key_chain.go (hunk 3 above)
Comment (StephenButtolph) [reply vms/spchainvm/key_chain.go]: It's a bit of a pain to require logging, so I just returned the error instead.

File: vms/spchainvm/key_chain.go [hunk 4]
```diff
@@ -28,16 +28,16 @@ type KeyChain struct {
 	Keys   []*crypto.PrivateKeySECP256K1R
```
[the remark this answers is not in the document]
Comment (StephenButtolph) [reply vms/spchainvm/key_chain.go]: Updated

File: xputtest/avm.go [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+	"github.com/ava-labs/gecko/utils/crypto"
+	"github.com/ava-labs/gecko/utils/formatting"
+	"github.com/ava-labs/gecko/utils/timer"
+	"github.com/ava-labs/gecko/vms/avm"
+	"github.com/ava-labs/gecko/vms/platformvm"
+	"github.com/ava-labs/gecko/xputtest/avmwallet"
+)
+
+func (n *network) benchmarkAVM(genesisState *platformvm.Genesis) {
+	avmChain := genesisState.Chains[0]
```
[the remark this answers is not in the document]
Comment (StephenButtolph) [reply xputtest/avm.go]: Updated

File: xputtest/avm.go [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+	n.log.AssertTrue(avmChain.ChainName == "AVM", "wrong chain name")
```
[the remark this answers is not in the document]
Comment (StephenButtolph) [reply xputtest/avm.go]: Updated

File: xputtest/avmwallet/wallet_test.go [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+	err = platformvm.Codec.Unmarshal(platformGenesisBytes, genesisState)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if err := genesisState.Initialize(); err != nil {
+		t.Fatal(err)
+	}
+
+	avmChain := genesisState.Chains[0]
+	if name := avmChain.ChainName; name != "AVM" {
```
[the remark this answers is not in the document]
Comment (StephenButtolph) [reply xputtest/avmwallet/wallet_test.go]: Updated

File: xputtest/chainwallet/wallet.go [hunk 8]
```diff
@@ -61,35 +55,40 @@ func (w *Wallet) Balance() uint64 { return w.balance }
...
+			}
+			w.accountSet[accountID.Key()] = newAccount
+			w.txs[i] = tx
+			i++
 		}
 	}
+	return nil
 }
 
 /*
```
[the remark this answers is not in the document]
Comment (StephenButtolph) [reply xputtest/chainwallet/wallet.go]: Incorporated it into the generate Txs function.

File: xputtest/dagwallet/utxo_set.go [hunk 9]
```diff
@@ -11,64 +11,64 @@ import (
 	"github.com/ava-labs/gecko/vms/spdagvm"
 )
 
-// UtxoSet ...
-type UtxoSet struct {
+// UTXOSet ...
+type UTXOSet struct {
 	// This can be used to iterate over. However, it should not be modified externally.
 	utxoMap map[[32]byte]int
```
[the remark this answers is not in the document]
Comment (StephenButtolph) [reply xputtest/dagwallet/utxo_set.go]: Updated

File: xputtest/avmwallet/wallet_test.go [hunk 10]
```diff
@@ -0,0 +1,315 @@
...
+	}
+	tx.Initialize(txBytes)
+
+	for _, utxo := range tx.UTXOs() {
+		w.AddUTXO(utxo)
+	}
+
+	assetID := genesisTx.ID()
+
+	if balance := w.Balance(assetID); balance != 45*units.MegaAva {
```
[the remark this answers is not in the document]
Comment (StephenButtolph) [reply xputtest/avmwallet/wallet_test.go]: This was meant to simulate running the xput test, so I changed the check to be less restrictive

File: xputtest/avm.go [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+	}
+
+	assetID := genesisTx.ID()
+
+	n.log.AssertNoError(wallet.GenerateTxs(config.NumTxs, assetID))
+
+	go n.log.RecoverAndPanic(func() { n.IssueAVM(avmChain.ID(), assetID, wallet) })
+}
+
+func (n *network) IssueAVM(chainID ids.ID, assetID ids.ID, wallet *avmwallet.Wallet) {
```
[the remark this answers is not in the document]
Comment (StephenButtolph) [reply xputtest/avm.go]: added some more comments

File: xputtest/main.go [hunk 12]
```diff
@@ -3,75 +3,22 @@
...
-	}
-
-	txID, err := ids.ToID(pMsg.Get(networking.TxID).([]byte))
-	t.log.AssertNoError(err) // Length is checked in message parsing
-
-	t.log.Debug("Decided %s", txID)
-	t.decided <- txID
-}
-
 func main() {
```
[the remark this answers is not in the document]
Comment (StephenButtolph) [reply xputtest/main.go]: Updated

**Comment by danlaine** [comment]
@StephenButtolph Can you add instructions for running to the readme? Once I run it I'll approve and add the same instructions to the docs.

**Comment by danlaine** [comment]
@StephenButtolph Hey, ran a throughput test locally.

One more ask: while the transactions are being generated, there is no message or anything so I'm afraid people will think the test just froze. Let's add prints saying `generated X of Y transactions to be sent in test` or something.

**Comment by danlaine** [comment]
One more thing, actually: When the test is done, it prints `done with test` but doesn't exit. Could you please make the program exit after this rather than the user having to `CTRL + C` to exit? Thanks!

**Pull request merged**